internal/handlers/pg/pgdb/query.go

```diff
@@ -246,19 +247,50 @@ func prepareWhereClause(sqlFilters *types.Document) (string, []any) {
 	var p Placeholder
 
 	for k, v := range sqlFilters.Map() {
-		switch k {
-		case "_id":
-			switch v := v.(type) {
-			case string:
-				filters = append(filters, fmt.Sprintf(`((_jsonb->'_id')::jsonb = %s)`, p.Next()))
-				args = append(args, string(must.NotFail(pjson.MarshalSingleValue(v))))
+		if len(k) != 0 && (k[0] == '$' || types.NewPathFromString(k).Len() > 1) {
+			// TODO $eq and $ne https://github.com/FerretDB/FerretDB/issues/1840
+			// TODO dot notation https://github.com/FerretDB/FerretDB/issues/1841
+			continue
+		}
 
-			case types.ObjectID:
+		// don't iterate through array for _id keys to simplify the query
+		if k == "_id" {
+			switch v := v.(type) {
+			case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
+				// type not supported for pushdown
+			case float64, string, types.ObjectID, int32, int64:
 				filters = append(filters, fmt.Sprintf(`((_jsonb->'_id')::jsonb = %s)`, p.Next()))
 				args = append(args, string(must.NotFail(pjson.MarshalSingleValue(v))))
+			default:
+				panic(fmt.Sprintf("Unexpected type of value: %v", v))
 			}
-		default:
+
+			continue
+		}
+
+		switch v := v.(type) {
+		case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
+			// type not supported for pushdown
 			continue
+
+		case float64, string, types.ObjectID, int32, int64:
+			sql := fmt.Sprintf(
+				// Select if value under the key k is equal to value v.
+				`((_jsonb->%[1]s)::jsonb = %[2]s)`+
+					// If it's not, but the value under the key k is an array - select if it contains the value equal to v.
+					` OR (_jsonb->%[1]s)::jsonb @> %[2]s`,
```

Comment:
containment operator "@>" while it's working on jsonb values will check if the value on the left contain at the top level the value on the right. If the left value (let's say field 'v') is an array it will check if it contains the value from the right, if it's the jsonb document like `{v:{"foo":42,"boo":32}}` then it will check if the value on the right is present inside it. 

If the value on the left is the jsonb value like '"foo"' or '42', then the @> comparator will compare it with the value on the right, it doesn't result in any error, so *technically* it's possible to simplify the where clause to just `WHERE (_jsonb->%[1]s)::jsonb @> %[2]s`, although I wouldn't do that, as it's not the proper usage of the operator, we're not sure how this comparison behaves in edge cases, and it's harder to understand when you read the code.

Reply:
> If the left value is an array
> If the left value is a jsonb document
> However, if the left value is a jsonb number

Please also consider strings. For example, how that would work given a document `{v: "foo"}` and filter `{v: "fo"}`, and vice versa?

Reply:
Checked that after the call with @chilagrow 

```SQL
select _jsonb from test."Composites_83e0b565" where (_jsonb->'v'->'42')::jsonb @> '"foo"';
```

While this query has some results, as we have 2 documents with `v.42 = "foo"` in collection

```SQL
select _jsonb from test."Composites_83e0b565" where (_jsonb->'v'->'42')::jsonb @> '"f"';
```

this one doesn't return anything